contracts/adapters/swap/source/src/contract.rs

```diff
@@ -0,0 +1,389 @@
+            &query_simulate_swap_exact_asset_out_with_metadata(
+                deps,
+                asset_out,
+                swap_operations,
+                include_spot_price,
+            )?,
+        )?),
+        QueryMsg::SimulateSwapExactAssetInWithMetadata {
+            asset_in,
+            swap_operations,
+            include_spot_price,
+        } => Ok(to_json_binary(
+            &query_simulate_swap_exact_asset_in_with_metadata(
+                deps,
+                asset_in,
+                swap_operations,
+                include_spot_price,
+            )?,
+        )?),
+        QueryMsg::SimulateSmartSwapExactAssetIn { asset_in, routes } => {
+            let ask_denom = get_ask_denom_for_routes(&routes)?;
+            Ok(to_json_binary(&query_simulate_smart_swap_exact_asset_in(
+                deps, asset_in, ask_denom, &routes,
+            )?)?)
+        }
+        QueryMsg::SimulateSmartSwapExactAssetInWithMetadata {
+            asset_in,
+            routes,
+            include_spot_price,
+        } => {
+            let ask_denom = get_ask_denom_for_routes(&routes)?;
+
+            Ok(to_json_binary(
+                &query_simulate_smart_swap_exact_asset_in_with_metadata(
+                    deps,
+                    asset_in,
+                    ask_denom,
+                    routes,
+                    include_spot_price,
+                )?,
+            )?)
+        }
+    }
+}
+
+fn query_simulate_swap_exact_asset_out(
+    deps: Deps,
+    asset_out: Asset,
+    operations: Vec<SwapOperation>,
+) -> Result<Asset, ContractError> {
+    query_simulate_swap_exact_asset_out_with_metadata(deps, asset_out, operations, false)
+        .map(|r| r.asset_in)
+}
+
+fn query_simulate_swap_exact_asset_out_with_metadata(
+    deps: Deps,
+    asset_out: Asset,
+    mut operations: Vec<SwapOperation>,
+    include_price: bool,
+) -> Result<SimulateSwapExactAssetOutResponse, ContractError> {
+    let coin_out = get_coin_from_asset(asset_out)?;
+    let (first_op, last_op) = (
+        operations
+            .first()
+            .ok_or(ContractError::SwapOperationsEmpty)?,
+        operations
+            .last()
+            .ok_or(ContractError::SwapOperationsEmpty)?,
+    );
+
+    // check that coin out matches the last swap operations out
+    if coin_out.denom != last_op.denom_out {
+        return Err(ContractError::CoinOutDenomMismatch);
+    }
+
+    let denom_in = first_op.denom_in.clone();
+
+    let router = ROUTER_CONTRACT_ADDRESS.load(deps.storage)?;
+
+    let first_hop = Hop {
+        pool: first_op.pool.clone(),
+        denom: first_op.denom_in.clone(),
+    };
+
+    let mut hops = vec![first_hop];
+    // we omit the last swap operation
+    // since that is already the coin_out
+    operations.pop();
+    // if it is not empty, we fill the other left operations.
+    if !operations.is_empty() {
+        hops.extend(operations.into_iter().map(|op| Hop {
```

Comment:
Doesn't this make the hops vector have the first hop twice? Since the first_hop is added at initialization of hops and then his extended here ?

Reply:
@NotJeremyLiu I also had some confusions releted to this so could you help me understand and what would be the better approach or current one is  fine.